### PR TITLE
Fix runtime check

### PR DESCRIPTION
**Pull request by borisfom**
# What does this PR do ?
Post-export runtime check refinements - allow checking multiple input sizes.
Restored max_dim argument functionality in export.py
Also, added a cast to ensure ASR nets are always exported with output in FP32 - even with autocast.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/models/asr_model.py [hunk 1]
```diff
@@ -125,6 +126,8 @@ def forward_for_export(self, input, length=None):
         else:
             decoder_input = encoder_output
         if hasattr(self.output_module, 'forward_for_export'):
-            return self.output_module.forward_for_export(decoder_input)
+            ret = self.output_module.forward_for_export(decoder_input)
         else:
-            return self.output_module(decoder_input)
+            ret = self.output_module(decoder_input)
+        # convert all FP16 results to FP32 for consistency
+        return cast_all(ret, from_dtype=torch.float16, to_dtype=torch.float)
```
Comment (titu1994) [inline nemo/collections/asr/models/asr_model.py:133]: Isn't torch.float = torch.float64 by default ?

File: nemo/utils/export_utils.py [hunk 2]
```diff
@@ -44,6 +44,23 @@ class ExportFormat(Enum):
 }
 
 
+def cast_tensor(x, from_dtype=torch.float16, to_dtype=torch.float):
+    return x.to(dtype=to_dtype) if x.dtype == from_dtype else x
+
+
+def cast_all(x, from_dtype=torch.float16, to_dtype=torch.float):
```
Comment (titu1994) [inline nemo/utils/export_utils.py:51]: Same comment as above. Should be explicit to fp32?

File: nemo/utils/export_utils.py [hunk 3]
```diff
@@ -44,6 +44,23 @@ class ExportFormat(Enum):
...
+    if isinstance(x, torch.Tensor):
+        return cast_tensor(x, from_dtype=from_dtype, to_dtype=to_dtype)
+    else:
+        if isinstance(x, dict):
+            new_dict = {}
+            for k in x.keys():
+                new_dict[k] = cast_all(x[k], from_dtype=torch.float16, to_dtype=torch.float)
+            return new_dict
```
Comment (titu1994) [inline nemo/utils/export_utils.py:59]: This should pass the user provided values, not the explicit defaults.

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall looks good, minor fix needed regarding recursive call and dtype

File: nemo/collections/asr/models/asr_model.py (hunk 1 above)
Comment (borisfom) [reply nemo/collections/asr/models/asr_model.py:133]: Nope. But sure let's use float32 here.

**Review by titu1994 — APPROVED** [review]
Looks fantaastic

**Pull request merged**